### PR TITLE
Add more Playwright tests

### PR DESCRIPTION
**Pull request by dabreegster**
See code comments for details. Biggest issues is that the tests are flaking on me locally, by timing out. #164 has ideas for more robustly waiting for maplibre to be ready, which could solve some problems. But even with proper waiting, running them is slow, because we use a totally new browser context for every single test. Nice for isolation, but slow. Pete and I are talking about trying [hooks](https://playwright.dev/docs/api/class-test#test-before-all) or fixtures to clear all state in between tests.

### CODE REVIEW

File: tests/modes.spec.ts [hunk 1]
```diff
@@ -1,19 +1,80 @@
-import { readFile } from "fs/promises";
 import { test, expect } from "@playwright/test";
 
-test("clearing all while a feature is open works", async ({ page }) => {
-  await page.goto("/scheme.html?authority=North Somerset");
-  await page.getByRole("button", { name: "New point" }).click();
+async function clickMap(page, x, y) {
```
Comment (dabreegster) [inline tests/modes.spec.ts:3]: Just a helper to reduce visual line noise below

File: tests/modes.spec.ts [hunk 2]
```diff
@@ -1,19 +1,80 @@
...
-    position: {
-      x: 500,
-      y: 500,
-    },
+    position: { x, y },
   });
+}
+
+test("clearing all while a feature is open works", async ({ page }) => {
+  await page.goto("/scheme.html?authority=Adur");
```
Comment (dabreegster) [inline tests/modes.spec.ts:10]: Switching to an authority with one of the smallest route files to load, in the hopes that it helps test speed

File: tests/modes.spec.ts [hunk 3]
```diff
@@ -1,19 +1,80 @@
...
+  await clickMap(page, 500, 500);
   page.on("dialog", (dialog) => dialog.accept());
   await page.getByRole("button", { name: "Clear all" }).click();
   // Make sure AttributeMode resets to nothing being selected
   await expect(
     page.getByText("Click an object to fill out its attributes")
   ).toBeVisible();
 });
+
+test("creating a new point opens a form", async ({ page }) => {
```
Comment (dabreegster) [inline tests/modes.spec.ts:21]: This should catch the problems that we uncovered in #196

File: tests/modes.spec.ts [hunk 4]
```diff
@@ -1,19 +1,80 @@
...
+    route.fulfill({
+      status: 404,
+    })
+  );
+  await page.goto("/scheme.html?authority=Adur");
+
+  await page.getByRole("button", { name: "New route" }).click();
+  await expect(page.getByText("Failed to load")).toBeVisible();
+
+  // Other tools should still work
```
Comment (dabreegster) [inline tests/modes.spec.ts:76]: @robinlovelace-ate, actually the point tool still works when route tool fails. Editing geometry might be broken though. I can't tell from your screenshot what happened exactly, but I'll keep playing with it and add tests to make sure other tools still work

File: tests/modes.spec.ts [hunk 5]
```diff
@@ -1,19 +1,119 @@
...
+  await page.getByRole("button", { name: "New point" }).click();
+  await clickMap(page, 500, 500);
+  await page.getByRole("button", { name: "1) Untitled point" }).isVisible();
+});
+
+test("editing geometry of a polygon works", async ({ page }) => {
```
Comment (dabreegster) [inline tests/modes.spec.ts:93]: For #218, except the test passes! Maybe it's flaky and indicates a race condition

File: tests/modes.spec.ts [hunk 6]
```diff
@@ -1,25 +1,132 @@
 import { readFile } from "fs/promises";
 import { test, expect, type Page } from "@playwright/test";
-import { loadInitialPage } from "./shared.ts";
+import {
+  loadInitialPage,
+  clickMap,
+  clearExistingInterventions,
+} from "./shared.ts";
```
Comment (dabreegster) [inline tests/modes.spec.ts:7]: Fun new config problem: `npm run check` says `Error: An import path can only end with a '.ts' extension when 'allowImportingTsExtensions' is enabled.`

I tried removing the `.ts` to be consistent with the rest of our code, but then `npm run test` fails. Apparently a different tsconfig is being applied for tests?

File: tests/modes.spec.ts [hunk 7]
```diff
@@ -1,25 +1,132 @@
...
 
 let page: Page;
 
 test.beforeAll(async ({ browser }) => {
   page = await loadInitialPage(browser);
 });
+test.beforeEach(async () => {
+  await clearExistingInterventions(page);
```
Comment (dabreegster) [inline tests/modes.spec.ts:15]: I forgot this at first and was very confused why tests failed together, but not individually!

File: tests/modes.spec.ts [hunk 8]
```diff
@@ -1,25 +1,132 @@
...
+  await page.getByLabel("Description:").click();
+
+  // Edit geometry, and click the polygon again
+  await page.getByRole("button", { name: "Edit geometry" }).click();
+  await clickMap(page, 312, 501);
+  // This button indicates the mode is working
+  await page.getByRole("button", { name: "Finish" }).click();
+});
+
+test("testing adding interventions, then deleting one, then adding another", async () => {
```
Comment (dabreegster) [inline tests/modes.spec.ts:104]: These two came from the deleted files. Moved them here for better grouping of mode/drawing tests, and to share the worker

File: tests/modes.spec.ts (hunk 4 above)
Comment (robinlovelace-ate) [reply tests/modes.spec.ts]: Will test...

Comment (robinlovelace-ate) [reply tests/modes.spec.ts]: I'm pretty sure I could not add a point or anything to download as a GeoJSON in the accidental test with no network.

Comment (dabreegster) [reply tests/modes.spec.ts]: https://acteng.github.io/atip/fail_route_load/scheme.html?authority=Adur
Give it ~10 mins to deploy, then we can play with this. It's set up to grab the wrong route files and fail, so we can see everything that breaks. I added a test (now in `errors.spec.ts` that demonstrates making a new point still works, at least)

Comment (robinlovelace-ate) [reply tests/modes.spec.ts]: @dabreegster cannot see how to add anything here.. https://acteng.github.io/atip/fail_route_load/scheme.html?authority=Adur#5.74/52.469/-1.707

Comment (dabreegster) [reply tests/modes.spec.ts]: Ahh. "New point" and freehand polygon work fine. But as soon as you try to click "New route", everything gets stuck. Thanks, we can update the tests accordingly and fix this bug

**Comment by dabreegster** [comment]
Oops, GH didn't add reviewers -- @Pete-Y-CS, how do these look to you? Rebased after the new hooks

File: tests/modes.spec.ts (hunk 6 above)
Comment (Pete-Y-CS) [reply tests/modes.spec.ts:7]: I'm not sure, I do remember   having some similar problem and just doing what got the tests to run and forgetting about it which might not be the best policy. I don't mind permitting importing with extensions but would be ideal to understand why this needs it and see if we can avoid

**Review by Pete-Y-CS — APPROVED** [review]

**Comment by dabreegster** [comment]
Thanks for the review! I'll merge this in now, so we can have some coverage for #220

**Pull request merged**